server/src/test/java/org/elasticsearch/cluster/coordination/AtomicRegisterCoordinatorTests.java

```diff
@@ -0,0 +1,693 @@
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+            PersistentClusterState that = (PersistentClusterState) o;
+            return term == that.term
+                && version == that.version
+                && Objects.equals(state.coordinationMetadata(), that.state().coordinationMetadata());
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(term, version, state.coordinationMetadata());
+        }
+    }
+
+    record TermOwner(DiscoveryNode node, long term) {
+        static TermOwner EMPTY = new TermOwner(null, 0);
+    }
+
+    static class KeyedAtomicRegister {
+        private final Map<String, PersistentClusterState> clusterStates = new HashMap<>();
+        private final Map<String, HeartBeat> heartBeats = new HashMap<>();
+        private final Map<String, TermOwner> currentTermOwners = new HashMap<>();
+
+        PersistentClusterState readClusterState(String key) {
+            return clusterStates.get(key);
+        }
+
+        boolean compareAndSetClusterState(String key, PersistentClusterState expectedState, PersistentClusterState newState) {
+            final var currentPersistentState = clusterStates.get(key);
+
+            final var winner = getTermOwner(key);
+            if (winner.term() > newState.term()) {
+                throw new RuntimeException(
+                    "There's a new leader in the cluster with term "
+                        + winner.term()
+                        + " while the cluster state update is for term "
+                        + newState.term()
+                );
+            }
+
+            if (Objects.equals(expectedState, currentPersistentState)) {
+                clusterStates.put(key, newState);
+                return true;
+            } else {
+                return false;
+            }
+        }
+
+        private TermOwner getTermOwner(String key) {
+            return currentTermOwners.getOrDefault(key, TermOwner.EMPTY);
+        }
+
+        boolean claimTerm(String key, DiscoveryNode newOwner, long newTerm) {
+            final var currentTermOwner = getTermOwner(key);
+
+            if (currentTermOwner.term() >= newTerm) {
+                return false;
+            }
+
+            currentTermOwners.put(key, new TermOwner(newOwner, newTerm));
+            return true;
+        }
+
+        void setHeartBeat(String key, HeartBeat heartBeat) {
+            heartBeats.put(key, heartBeat);
+        }
+
+        @Nullable
+        HeartBeat latestHeartBeat(String key) {
+            return heartBeats.get(key);
+        }
+
+        void removeHeartBeat(String key, HeartBeat latestHeartBeat) {
+            heartBeats.remove(key, latestHeartBeat);
+        }
+    }
+
+    class AtomicRegisterPersistedState implements CoordinationState.PersistedState {
```

Comment:
Now that we have the one-leader-per-term property, do we need the state persistence itself to be register-based too? I think the state will only be written by the (unique) leader of the corresponding term, so as long as we include the term in the blob container path we don't need to be so strict here.

(I expect this means that a newly-elected leader may need to search through past terms to find the latest state instead)

Reply:
I've moved the cluster state storage into a shared store in the latest commits, together with heartbeats as you suggested.